### ISSUE TITLE
How to use gulp-mem as src ?

### ISSUE THREAD

**Issue by MickL** [turtlemay/gulp-mem#4]
I can write files into gulp-mem. But how can i use this files as `src` again?

```
gulp.task('webpack', () => {
    return gulp
        .src(global.gulpMem.???('js/main.js'))
        .pipe(webpack())
        .pipe(global.gulpMem.dest('.built/js/'))
    ;
});
```

**Comment by turtlemay** [comment]
Why the `global`? It looks like you're trying to persist data between task runs, but that's not possible because they don't share the same context. (Not without using the actual filesystem, which is what we're avoiding in the first place.) The way to continuously interact with an in-memory object from multiple processes is through a server.

I wasn't supporting `src` because the mock filesystem only exists in the running process and contains nothing until you pipe into it, so I didn't see a reason to use it at the start of a sequence like that.

Although, I suppose populating the filesystem yourself before starting a sequence might be a valid use case. I'll think about that.

**Comment by MickL** [comment]
gulp.watch runs a sequence of tasks. But im not able to access the source files anymore.

**Comment by turtlemay** [comment]
This is the currently supported usage:

```js
gulp.task('start-dev', () => {
  // Start dev server and use the middleware to
  // serve your built files from memory...
  // ...

  // Rebuild after source file changes.
  // The dev server will continually serve the updated files.
  gulp.watch('./src/**/*', ['build:dev'])
})
```

What do you need to do with the in-memory files after you've built them that isn't part of your build task?

**Comment by saborrie** [comment]
You can access the contents of the files via the `fs` property on your `gulpMem`. You can't exactly use it the same way as `gulp.src`;  you would have to create your own stream and then put the contents of `gulpMem.fs.readFileSync('js/main.js')` in there.

**Comment by saborrie** [comment]
It's probably worth pointing out that MemoryFS (the in memory filesystem that this library uses), is written by webpack and is used by webpack-dev-server. If you're building using webpack, it's probably better to just use that instead of gulp.

**Comment by MickL** [comment]
E.g. gulp-server-livereload is pretty fast but i have to declare a src:

```
const server = require('gulp-server-livereload');

gulp.src(global.gulpMem.???)
                .pipe(server({
                    livereload: {
                        enable: true
                    },
                }));
```

**Comment by MickL** [comment]
@turtlemay @saborrie e.g. gulp-server-livereload is pretty fast but i have to declare a src:

```
const server = require('gulp-server-livereload');

gulp.src(global.gulpMem.???)
    .pipe(server({
        livereload: {
            enable: true
        },
    }))
;
```

**Comment by saborrie** [comment]
I see what you're saying, perhaps there is a reasonable expectation to be able to use gulp-mem for src as well as dest.

In any case, to solve your specific example, you can use the middleware. Other common devserver modules such as connect and browsersync support express-style middleware properly, but gulp-server-livereload doesn't. It only supports middleware for fallback files, and it appears the fallback file must actually exist.

Try the following:

Folder structure:
```
.
├───dist
│   └───fromdist.html
│
├───node_modules
│
├───src
│   └───fromsrc.html
│
├───gulpfile.js
│
└───package.json
```

Gulpfile:
```js
const fs = require('fs')
const gulp = require('gulp')
const Mem = require('gulp-mem')
const server = require('gulp-server-livereload')

const mem = new Mem()
mem.serveBasePath = './dist'

gulp.task('copy', function () {
  return gulp.src('src/**/*.html')
    .pipe(mem.dest('./dist'))
})

gulp.task('webserver', ['copy'], function() {
  gulp.src('./dist')
    .pipe(server({
      fallback: 'fromdist.html', //this must actually exist inside ./dist
      fallbackLogic: function(req, res, fallbackFile) {
        mem.middleware(req, res, function () {
          fs.createReadStream(fallbackFile).pipe(res)
        })
      }
    }))
})

gulp.task('default', ['webserver'])
```

if you try the above you'll be able to browse to 'localhost:8000/fromsrc.html' and see the file from the src directory served via memoryfs

**Comment by MickL** [comment]
Thats not working because ./dist doesn't exists when writing in memory. Also im not sure if it only serves fromdist.html and nothing more?

Unfortionally BrowserSync is not an option since css-injection doesn't work(reloads the page) and gulp-connect requires an additional browser plugin. gulp-server-livereload was the best i found so far: Using socket.io its blazing fast.

**Comment by saborrie** [comment]
Yeah, the `dist/fromdist.html` must exist. You could just make it a 404 page I guess.

I took a look at the code from gulp-server-livereload and I'm not convinced implementing gulpMem.src would help with what you want to do. It looks like gulp-server-livereload accesses the filesystem itself via node-watch in order to reload files, which would bypass any attempt to pipe files from a different filesystem like memoryFs. Essentially the original piped copy of the files could work, but node-watch would never detect changes to those files, so they would never be reloaded.

You could try forking gulp-server-reload and changing its api to allow custom middleware to be passed in, but it is probably a very large task.
